### PR TITLE
Add first paint, first contentful paint metrics

### PR DESCRIPTION
**Pull request by zephraph**
## Changes

- Add separate user performance timing api
- Simplify volley api to reduce duplicate logic

## Todo

- [x] Update volley test to support new api
- [x] Fix ignored ts errors
- [x] Add comments

### CODE REVIEW

File: src/lib/user-performance-metrics.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      perf.measure(measureName, startMarkName)
+    } catch (err) {
+      console.error(err)
+    }
+  }
+}
+
+// default metrics
+
+// export function getEffectiveConnectionType() {
```
Comment (zephraph) [inline src/lib/user-performance-metrics.ts]: A few of these metrics (this one and getDeviceMemory) might not really be needed. I may just go ahead and remove them.

File: src/lib/user-performance-metrics.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+//     typeof navigator !== "undefined"
+//       ? navigator.connection ||
+//         navigator.mozConnection ||
+//         navigator.webkitConnection
+//       : null
+//   return conn ? conn.effectiveType : null
+// }
+
+export function getFirstPaint() {
+  if (typeof PerformancePaintTiming === "undefined") return null
```
Comment (alloy) [inline src/lib/user-performance-metrics.ts]: Do we really need to test `PerformancePaintTiming` instead of `perf !== null`?

File: src/lib/user-performance-metrics.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  if (perf && perf.mark) {
+    perf.mark(markName)
+  }
+}
+
+export function measure(measureName, startMarkName) {
+  if (perf && perf.measure) {
+    try {
+      perf.measure(measureName, startMarkName)
+    } catch (err) {
```
Comment (alloy) [inline src/lib/user-performance-metrics.ts]: Is there any specific exceptions that `measure` may throw? If so, can you maybe add a comment that describes them or a link to a definitive doc?

File: src/lib/user-performance-metrics.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  const firstContentfulPaint = perf
+    .getEntriesByType("paint")
+    .find(({ name }) => name === "first-contentful-paint")
+  return firstContentfulPaint
+    ? Math.round(firstContentfulPaint.startTime)
+    : null
+}
+
+export function getOnLoad() {
+  if (!perf || !perf.timing) return null
```
Comment (alloy) [inline src/lib/user-performance-metrics.ts]: Maybe define a ‘timing available’ const?

File: src/lib/user-performance-metrics.ts (hunk 1 above)
Comment (alloy) [reply src/lib/user-performance-metrics.ts]: Hmm, yeah what could we learn from the device’s memory pressure that would be helpful?

File: src/lib/user-performance-metrics.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      name: measure.name,
+      startTime: Math.round(measure.startTime),
+      duration: Math.round(measure.duration),
+    }
+  })
+  return marks.concat(measures)
+}
+
+export function getResources() {
+  if (!perf || typeof PerformanceResourceTiming === "undefined") return null
```
Comment (alloy) [inline src/lib/user-performance-metrics.ts]: Is this different from `!perf || !perf.timing` ?

File: src/lib/user-performance-metrics.ts (hunk 2 above)
Comment (zephraph) [reply src/lib/user-performance-metrics.ts]: Yeah. That's a feature flag for that specific measurement. If it isn't present then there will be no paint events captured and if it is we know the performance interface exists because it's a subset of that system.

Comment (alloy) [reply src/lib/user-performance-metrics.ts]: Ah gotcha 👍

File: src/lib/user-performance-metrics.ts (hunk 5 above)
Comment (zephraph) [reply src/lib/user-performance-metrics.ts]: Yeah. It's more specific. Technically the `!perf` check doesn't need to be performed her because if `perf` is null then `PerformanceResourceTiming` will absolutely be null. Still, as with `PerformancePaintTiming`, `PerformanceResourceTiming` is only present when that specific performance measuring feature is present. I'm actually just going to remove this though, because it's not needed right now.

File: src/lib/user-performance-metrics.ts (hunk 3 above)
Comment (zephraph) [reply src/lib/user-performance-metrics.ts]: Yeah, I can add some comments here. If the markNames specified don't exist (i.e. the mark hasn't been created) then it'll throw an exception.

File: src/lib/user-performance-metrics.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
+// Forked and modified from UXM: https://github.com/treosh/uxm
```
Comment (damassi) [inline src/lib/user-performance-metrics.ts]: New file names are typically `camelCase` (where old file names were `snake_case`)

File: src/lib/user-performance-metrics.ts (hunk 3 above)
Comment (alloy) [reply src/lib/user-performance-metrics.ts]: Better yet, maybe add a test that describes that situation?

Comment (zephraph) [reply src/lib/user-performance-metrics.ts]: The test would be a little harder because it would need to run in a browser environment to actually be reliable. I'll add a comment though.

File: src/lib/user-performance-metrics.ts (hunk 6 above)
Comment (alloy) [reply src/lib/user-performance-metrics.ts]: Thanks!

**Comment by alloy** [comment]
Awesome, let’s do this 👍
